design-proposals/klip-52-bytes-data-type-support.md

```diff
@@ -0,0 +1,156 @@
+# KLIP 46 - BYTES Data Type Support
+
+**Author**: Zara Lim (@jzaralim) | 
+**Release Target**: 0.21 | 
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7764
+
+**tl;dr:** _Add support for the BYTES data type. This will allow users to work with BLOBs of data that don't fit into any other data type._
+           
+## Motivation and background
+
+Currently, ksqlDB can only handle a set of primitive types and combinations of them.
+A BYTES data type would allow users to work with data that does not fit into any of
+the primitive types such as images, as well as BLOB/binary data from other databases.
+
+## What is in scope
+* Add BYTES type to KSQL
+* Support BYTE comparisons
+* Support BYTES usage in STRUCT, MAP and ARRAY
+* Serialization and de-serialization of BYTES to Avro, JSON, Protobuf and Delimited formats
+* Adding/updating UDFs to support the BYTES type
+* Casting between BYTES and STRING
+
+## What is not in scope
+* Fixed sized BYTES (`BYTES(3)` representing 3 bytes, for example) - This is supported by Kafka Connect by adding the `connect.fixed.size`
+key in a bytes schema, but this will not be included in this KLIP.
+
+## Public APIS
+
+### BYTES
+
+The BYTES data type will store an array of raw bytes of an unspecified length. The maximum size of
+the array is limited by the maximum size of a Kafka message, as well as possibly by the value format being used.
+The syntax is as follows:
+
+```roomsql
+CREATE STREAM stream_name (b BYTES, COL2 STRING) AS ...
+CREATE TABLE table_name (col1 STRUCT<field BYTES>) AS ...
+```
+
+By default, BYTES will be displayed in console as HEX strings, where each byte is represented by two characters.
+For example, the byte array `[91, 67]` will be displayed as:
+
+```roomsql
+> SELECT b from STREAM;
+'0x5B43'
+```
+
+Users can also represent BYTES as HEX strings, for example
+
+```roomsql
+> INSERT INTO STREAM VALUES ('0x5b43', 'string value');
+```
+
+The input and output formats can be configured using a new property, `ksql.bytes.format`.
```

Comment:
What about letting users use functions instead of setting configs? It will be easier to identify how the data is displayed by seeing the schema with `describe ...` when using functions.

Reply:
That works. The only downside is, the output format will always be a HEX string, though the user can use a conversion function if they want to see another output format.